research/gnn-survey/layers.py

```diff
@@ -40,15 +42,75 @@ def build(self, input_shape):
     if self.bias:
       self.b = self.add_weight(
           name='bias',
-          shape=(self.output_dim,),
+          shape=(self.output_dim, ),
           initializer='random_normal',
           trainable=True)
 
   def call(self, inputs):
     x, adj = inputs[0], inputs[1]
-    x = tf.matmul(adj, x)
+    if self.sparse:
+      x = tf.sparse.sparse_dense_matmul(adj, x)
+    else:
+       x = tf.matmul(adj, x)
     outputs = tf.matmul(x, self.weight)
+
     if self.bias:
       return self.b + outputs
     else:
       return outputs
+
+class GraphAttnLayer(tf.keras.layers.Layer):
+  """ Single graph attention layer."""
+
+  def __init__(self, output_dim, dropout_rate, alpha=0.2, **kwargs):
+    """Initializes the GraphAttnLayer.
+
+    Args:
+      output_dim: (int) Output dimension of gat layer
+      dropout_rate: (float) Dropout probability
+      alpha: (float) LeakyReLU angle of alpha
+      **kwargs: Keyword arguments for tf.keras.layers.Layer.
+    """
+    super(GraphAttnLayer, self).__init__(**kwargs)
+    self.output_dim = output_dim
+    self.dropout_rate = dropout_rate
+    self.alpha = alpha
+
+    self.leakyrelu = tf.keras.layers.LeakyReLU(self.alpha)
+    self.dropout = tf.keras.layers.Dropout(self.dropout_rate)
+
+  def build(self, input_shape):
+    super(GraphAttnLayer, self).build(input_shape)
+    self.weight = self.add_weight(
+        name='weight',
+        shape=(input_shape[0][-1], self.output_dim),
+        initializer='random_normal',
+        trainable=True)
+
+    self.attention = self.add_weight(
+        name='attention',
+        shape=(2*self.output_dim, 1),
+        initializer='random_normal',
+        trainable=True)
+
+
+  def call(self, inputs):
+    x, adj = inputs[0], inputs[1]
+    x = self.dropout(x)
+    x = tf.matmul(x, self.weight)
+
+    # feat shape: [n_data, n_data, output_dim]
+    n_data = 2708
```

Comment:
I don't think this should be hard-coded here? Can you get it from somewhere else?

Reply:
I just forgot to modify this to the latest commit, now I fixed it, sorry:(